`to_timestamp_millis()`, `to_timestamp_micros()`, `to_timestamp_seconds()`

velvia opened the pull request:
# Which issue does this PR close?

Closes #355 .

 # Rationale for this change

This PR implements the timestamp functions above in #355, basically extending the following capabilities for all Timestamp types (specifically for Millis, micros, and seconds) via the above named functions:

- Ability to convert timestamp strings to Timestamp type
- Ability to cast Int64 numeric arrays into equivalent Timestamp types

# What changes are included in this PR?

Pretty straightforward additions to `functions.rs`, `datetime_expressions` etc.

What is NOT included so far but could be depending on discussion:
- Not sure where to add documentation for these functions
- It might be useful to users to be able to cast different Timestamp types into others, for example, Timestamp-Nanos to Timestamp-Millis, and that would be straightforward to add as well.

# Are there any user-facing changes?

Three new user-facing SQL functions.

velvia commented:
/cc @alamb

alamb commented:
BTW I plan to review this later today

alamb commented on datafusion/src/physical_plan/expressions/cast.rs:
```diff
@@ -91,24 +91,26 @@ impl PhysicalExpr for CastExpr {
+}
+
+/// Internal cast function for casting ColumnarValue -> ColumnarValue for cast_type
+pub fn cast_array(
```
maybe this would be more clearly named `cast_column` or `cast_columnar_value` (as it might be casting an array or a Scalar)

alamb commented on datafusion/src/physical_plan/functions.rs:
```diff
@@ -1023,6 +1098,18 @@ fn signature(fun: &BuiltinScalarFunction) -> Signature {
         BuiltinScalarFunction::ToTimestamp => Signature::Uniform(1, vec![DataType::Utf8]),
+        BuiltinScalarFunction::ToTimestampMillis => Signature::Uniform(
+            1,
+            vec![DataType::Utf8, DataType::LargeUtf8, DataType::Int64],
```
These signatures claim to support arguments of type`LargeUtf8` but `LargeUtf8` isn't handled in the implementations above. I think we should make them consistent (I don't have a preference about actually supporting `LargeUtf`, btw)

velvia replied:
@alamb I actually don't understand what LargeUTF is about, but since `ToTimestamp` only supports regular Utf8, is it OK if all of the timestamp functions just support regular Utf8?

alamb replied:
> @alamb I actually don't understand what LargeUTF is about,

`Utf8`uses i32 to store offsets and thus can store strings up to 2GB.  `LargeUtf8`  stores string offsets using u64s and thus can support strings that are larger than 2GB. I do not know how widely `LargeUtf8` is used in practice. 

> is it OK if all of the timestamp functions just support regular Utf8? 

yes I think so

alamb commented on datafusion/tests/sql.rs:
```diff
@@ -2322,17 +2405,33 @@ async fn like() -> Result<()> {
+    let nanotimestamps = vec![
+        1599572549190855000i64, // 2020-09-08T13:42:29.190855+00:00
+        1599568949190855000,    // 2020-09-08T12:42:29.190855+00:00
+        1599565349190855000,
```
```suggestion 
        1599565349190855000, //2020-09-08T11:42:29.190855+00:00
```

alamb commented on datafusion/tests/sql.rs:
```diff
@@ -2322,17 +2405,33 @@ async fn like() -> Result<()> {
+        1599568949190855000,    // 2020-09-08T12:42:29.190855+00:00
+        1599565349190855000,
+    ]; // 2020-09-08T11:42:29.190855+00:00
+    let divisor = match A::get_time_unit() {
```
👍

alamb reviewed (approved):
Looks great to me. Thank you @velvia 

One thing I noticed while testing was that there wasn't a convenient way to cast a timestamp (nanoseconds) -- one had to go through in64. 

So for example I can't apply `to_timestamp_millis` to the output of `to_timestamp()`:
```
> select to_timestamp('2020-09-08T12:00:00+00:00')
> ;
+------------------------------------------+
| TimestampNanosecond(1599566400000000000) |
+------------------------------------------+
| 2020-09-08 12:00:00                      |
+------------------------------------------+
1 row in set. Query took 0.002 seconds.

> select to_timestamp_millis(to_timestamp('2020-09-08T12:00:00+00:00'));
Internal("Unsupported data type Ok(Timestamp(Nanosecond, None)) for function to_timestamp_millis")
```

However, there is a (very ugly) workaround using `cast`:

```
select to_timestamp_millis((cast(to_timestamp('2020-09-08T12:00:00+00:00') as bigint)) / 1000000);
+---------------------------------------------------------------------+
| totimestampmillis(Int64(1599566400000000000) Divide Int64(1000000)) |
+---------------------------------------------------------------------+
| 2020-09-08 12:00:00                                                 |
+---------------------------------------------------------------------+
```

velvia commented:
@alamb thanks for the review.

> One thing I noticed while testing was that there wasn't a convenient way to cast a timestamp (nanoseconds) -- one had to go through in64.

I think I'll go ahead and add support for the use case of `to_timestamp_xxx` casting from other timestamp columns.  I can imagine that being very useful, for us at least.  :)

Also, I had asked about docs - I think the `to_timestamp_xxx` functions should be documented somewhere, since they are nonstandard wrt most SQL specs.   The main README doesn't list explicitly supported timestamp functions, which we could, but I'm not sure anywhere else they could go.   There is `docs/specification`, and then there is a user guide which looks like it is in an early stage of progress.   Suggestions welcome!

alamb commented:
> Also, I had asked about docs - I think the to_timestamp_xxx functions should be documented somewhere, since they are nonstandard wrt most SQL specs. The main README doesn't list explicitly supported timestamp functions, which we could, but I'm not sure anywhere else they could go. There is docs/specification, and then there is a user guide which looks like it is in an early stage of progress. Suggestions welcome!

There is a list of supported functions here:
https://github.com/apache/arrow-datafusion#sql-support

Up to now, we have mostly been following the postgres functions and thus didn't need to document functions within DataFusion.

I suggest:
1.  start a section in the docs/userguide for DataFusion specific functions
2. Add entries for `to_timestamp_millis()` in the list at https://github.com/apache/arrow-datafusion#sql-support linked to the user guide

velvia commented:
I'm getting a very odd local compile error after merging with master latest:

```
error[E0432]: unresolved import `arrow::compute::kernels::partition`
  --> datafusion/src/physical_plan/mod.rs:28:30
   |
28 | use arrow::compute::kernels::partition::lexicographical_partition_ranges;
   |                              ^^^^^^^^^ could not find `partition` in `kernels`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `datafusion`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error[E0432]: unresolved import `arrow::compute::kernels::partition`
  --> datafusion/src/physical_plan/mod.rs:28:30
   |
28 | use arrow::compute::kernels::partition::lexicographical_partition_ranges;
   |                              ^^^^^^^^^ could not find `partition` in `kernels`

error: aborting due to previous error
```

alamb commented:
> I'm getting a very odd local compile error after merging with master latest:


@velvia  I think you need to do `cargo update` locally to pick up the latest release of arrow (4.3)

We should probably update the version requirements in datafusion to reflect that it requires arrow 4.3 or later (not 4.0 or later)

velvia commented:
@alamb thanks, I can update cargo.toml for datafusion, that did the trick

velvia commented:
I believe all changes and feedback have been addressed.  I added the capability of to_timestamp_xx() to convert from other timestamp types for convenience, and also added documentation.   :)

alamb commented on README.md:
```diff
@@ -320,31 +324,31 @@ execution. The SQL types from
+| `BOOLEAN`     | `Boolean`                         |
+| `DATE`        | `Date32`                          |
+| `TIME`        | `Time64(TimeUnit::Millisecond)`   |
+| `TIMESTAMP`   | `Timestamp(TimeUnit::Nanosecond)` |
```
👍

alamb commented on docs/user-guide/src/sql/datafusion-functions.md:
```diff
@@ -0,0 +1,86 @@
+
+## `to_timestamp`
+
+`to_timestamp()` is similar to the standard SQL function. It performs conversions to type `Timestamp(Nanoseconds, None)`, from:
```
Great docs

alamb commented:
Thanks @velvia  -- this looks great!
